### ISSUE TITLE
Exposing analytical nuclear gradient components

### ISSUE THREAD

**Issue by rlaplaza** [psi4/psi4#2130]
Hello! Thank you all for this amazing piece of code. 

I would just like to know if it is possible to obtain the analytical gradient decomposed. In particular, I would like to know if the call:

`wfn = procedures['gradient'][lowername](lowername, molecule=molecule, **kwargs)`

can be modified to get me the gradient per KS-MO, exchange-correlation, and nuclear repulsion in a KS-DFT calculation (ideally from the Python API). I think this could be useful for many purposes, e.g. testing analytical gradients for custom Wee terms of the Hamiltonian, for instance; and it just requires exposing it. 

Thanks in advance!

**Comment by hokru** [comment]
The readily available scf gradient terms can be found here: https://github.com/psi4/psi4/blob/master/psi4/src/psi4/scfgrad/scf_grad.cc#L122

Not sure about the best way to expose them.

**Comment by JonathonMisiewicz** [comment]
"Gradient per KS-MO" does _not_ "just require exposing it." First, you need to define it properly. Exact exchange depends on two KS-MOs, not just one, so it isn't clear to me how you'd meaningfully define this as a one KS-MO property. Then, you'd need to implement it, since Psi4 doesn't do this naturally. You'd probably need to decompose the density as a sum of densities for each MO and run the gradient computation on each, so a gradient computation takes #MO's times longer. This should not be the default algorithm.

Putting the gradient terms on the wavefunction as array variables seems relatively straightforward.

**Comment by hokru** [comment]
I assume it's about the paper from our(?) twitter conversation? https://onlinelibrary.wiley.com/doi/abs/10.1002/jcc.26459

**Comment by rlaplaza** [comment]
Indeed, @JonathonMisiewicz , the terms coming from the derivatives of two-electron integrals are not trivially available or easy to cast into one-body form. However, the total terms (Nuclear, Core, Coulomb...) should, and that would be enough to make me happy. I do think that accessing these terms can be useful for the userbase. 

Exposing the one-electron terms for Overlap etc. should also be possible, but this is both silly and shaky as a feature (i.e. not consistent across levels of theory) and sometimes a picky user should have to differentiate like a grown-up!

**Comment by JonathonMisiewicz** [comment]
Some comments for posterity:
* The Coulomb terms also come from the derivatives of two-electron integrals. You can (and for efficiency should) contract the TEI derivatives against the 1RDM to get the Coulomb terms. Conceptually, this is the interaction of an orbital with the mean field of all orbitals. It's not a pure one-electron interaction, but it's close enough.
* For purposes of thinking about the implementation, this issue is two very different ones:  (1) exposing the existing pieces of the SCF gradient and (2) further decomposing those pieces into per-orbital contributions, where applicable. (1) is as easy as setting array variables on the wavefunction. Per-orbital contributions aren't even computed, as they shouldn't be in an efficient code, so (2) is non-trivial.
